### ISSUE TITLE
Reorganize _pywt

### ISSUE THREAD

**Issue by kwohlfahrt** [PyWavelets/pywt#127]
In my opinion, the size of `_pywt.pyx` has gotten a bit out of hand (basically I'm having trouble finding things). I propose to split it up into smaller files, with the following organization:

```
pywt
|-dwt.py
| |-dwt
| |-idwt
| |-upcoef
| +-downcoef
|-swt.py
| +-swt
+-src
  |-c
  | +-*.[c,src]
  |-_dwt.pyx
  | |-dwt
  | |-dwt_axis
  | |-downcoef
  | |-idwt
  | |-idwt_axis
  | +-_upcoef
  +-_swt
    +-swt
```

This would also move most of the non-typed bits into pure python. I could keep them in Cython, but they're almost valid python as-is and there are no loops or other heavy bits in them.

Any thoughts, comments or should I get on with it?

**Comment by grlee77** [comment]
I am +1 on this.  I think it makes it a bit easier to find things and would also reduce the size of the cython-generated C code.

one minor point:
In #38, @aaren renamed the modules `multidim.py` to `_multidim.py`, etc, so I would follow that convention with `dwt.py`->`_dwt.py` and `swt.py`->`_swt.py`.

**Comment by rgommers** [comment]
+1 on both splitting up the Cython code and on adding underscores.

**Comment by aaren** [comment]
+1. I'd like to get #55 in first though to avoid possibly messy rebasing (should be in the next week).

**Comment by kwohlfahrt** [comment]
@grlee77 I think that would clash with imports for the cython modules, wouldn't it?

- `pywt/src/_dwt.pyx` -> `build/lib.xyz/pywt/_dwt.cpython-xyzm.so`
- `pywt/_dwt.py` -> `build/lib.xyz/pywt/_dwt.py`

**Comment by grlee77** [comment]
right.  perhaps rename `_dwt.pyx` to something like `_dwt_cy.pyx`.  There are a few places in `skimage` where that convention is used, but I don't know if it is a common one.

**Comment by kwohlfahrt** [comment]
What is the reasoning behind the rename to `_multidim.py`? I don't quite understand so perhaps that'll help me figure out what to do.

**Comment by aaren** [comment]
`multidim` -> `_multidim` is because the contents are not necessarily public API - this is indicated by the preceding `_`. API access is defined in `__init__.py`.

**Comment by kwohlfahrt** [comment]
I see. I think I'll put everything compiled in a sub-module (`_extensions`) for now. It might be time to think about organizing the namespace as a whole.

It would make sense to have some partitioning by transform (e.g. `dwt`, `swt`, `modwt`), and perhaps in each have a `multilevel` and `multidim` file? But I'm not sure how things that are both multilevel and multidimensional would fit into that. Hmmm.

**Comment by rgommers** [comment]
Keep in mind "flat is better than nested". Right now there are about ~50 objects in the ``pywt`` namespace. Partitioning that up isn't really needed I think. Maybe when we add new types of transforms though, like the curvelets that were proposed a while ago, those should get their own submodule.

**Comment by kwohlfahrt** [comment]
Fair enough. I like my functions organized into lots of small modules, so I'll leave the organization up to you guys.

**Comment by grlee77** [comment]
closed by #148
